ui/src/app/utils/cluster-initializer.tsx

```diff
@@ -0,0 +1,301 @@
+const DEFAULT_MAX_DELETE_COUNT = 2;
+const DEFAULT_MAX_RESUME_COUNT = 2;
+const DEFAULT_MAX_SERVER_ERROR_COUNT = 5;
+
+export class ClusterInitializationFailedError extends Error {
+  public readonly cluster: Cluster;
+
+  constructor(message: string, cluster?: Cluster) {
+    super(message);
+    Object.setPrototypeOf(this, ClusterInitializationFailedError.prototype);
+
+    this.name = 'ClusterInitializationFailedError';
+    this.cluster = cluster;
+  }
+}
+
+export class ExceededActionCountError extends ClusterInitializationFailedError {
+  constructor(message, cluster?: Cluster) {
+    super(message, cluster);
+    Object.setPrototypeOf(this, ExceededActionCountError.prototype);
+
+    this.name = 'ExceededActionCountError';
+  }
+}
+
+export class ExceededErrorCountError extends ClusterInitializationFailedError {
+  constructor(message, cluster?: Cluster) {
+    super(message, cluster);
+    Object.setPrototypeOf(this, ExceededErrorCountError.prototype);
+
+    this.name = 'ExceededErrorCountError';
+  }
+}
+
+export interface ClusterInitializerOptions {
+  // Core options. Most callers should provide these.
+  //
+  // The workspace namespace to initialize a cluster for.
+  workspaceNamespace: string;
+  // Callback which is called every time the cluster updates its status. This callback is *not*
+  // called when no cluster is found.
+  onStatusUpdate?: (ClusterStatus) => void;
+  // An optional abort signal which allows the caller to abort the initialization process, including
+  // cancelling any outstanding Ajax requests.
+  abortSignal?: AbortSignal;
+
+  // Override options. These options all have sensible defaults, but may be overridden for testing
+  // or special scenarios (such as an initialization flow which should not take any actions).
+  initialPollingDelay?: number;
+  maxPollingDelay?: number;
+  overallTimeout?: number;
+  maxCreateCount?: number;
+  maxDeleteCount?: number;
+  maxResumeCount?: number;
+  maxServerErrorCount?: number;
+}
+
+/**
+ * A controller class implementing client-side logic to initialize a Leonardo cluster. This class
+ * will continue to poll the getCluster endpoint, taking certain actions as required to nudge the
+ * cluster towards a running state, and will eventually resolve the Promise with a running cluster,
+ * or otherwise reject it with information about the failure mode.
+ *
+ * This is an unusually heavyweight controller class on the client side. It's worth noting a couple
+ * reasons why we ended up with this design:
+ *  - Cluster initialization can take up to 10 minutes, which is beyond the scope of a single
+ *    App Engine server-side request timeout. To reliably implement this control on the server side
+ *    would likely require new database persistence, tasks queues, and additional APIs in order to
+ *    provide the client with status updates.
+ *  - Ultimately, we might expect this type of functionality ("Get me a cluster for workspace X and
```

Comment:
I'm not sure I agree with this statement. We've conceptually merged several Leo concepts into one within the AoU product, primarily on account of product decisions to avoid exposing the inner workings of the cluster state. This creates some impedance mismatches which generates complexity here. For example, we don't really expose the fact that a cluster can become paused, or that it can become deleted. If we decided to surface these things (as Terra UI probably does), it would create less business logic here.

-  Polling cannot be removed, unless cluster creation is redesigned and made instantaneous somehow. Even if we pushed the maximum amount of logic into Leo, the operation is fundamentally asynchronous.
-  Logic around when to create/start/stop a cluster at the request of a user needs to live at the product level.
-  "recreate this cluster" could potentially be a Leo operation, but since we customize Leo clusters anyways, we probably want a hand in recreating clusters anyways.

The other avenue we could explore would be to say that Leo should not have an error state, but ultimately some cluster creations will fail regardless of how many retries Leo adds. So I do think a product-level cluster error state is still a necessity.

Reply:
Yeah, the less we know about the underlying clusters the better. We shouldn't know whether we have a new or existing cluster ideally (provided we can supply config).